ledger-core/virtual/execute/execute.go

```diff
@@ -223,36 +239,42 @@ func (s *SMExecute) stepWaitObjectReady(ctx smachine.ExecutionContext) smachine.
 			// attempt to create object that is deactivated :(
 			panic(throw.NotImplemented())
 		}
-	} else {
-		// if not constructor
-		switch objectState {
-		case object.Unknown:
-			panic(throw.Impossible())
-		case object.Missing:
-			s.prepareExecutionError(throw.E("object does not exist", struct {
-				ObjectReference string
-				State           object.State
-			}{
-				ObjectReference: s.execution.Object.String(),
-				State:           objectState,
-			}))
-			return ctx.Jump(s.stepSendCallResult)
-		case object.Inactive:
-			panic(throw.NotImplemented())
-		case object.HasState:
-			// ok
-		}
+
+		// default isolation for constructors
+		s.methodIsolation = contract.ConstructorIsolation()
+
+		return ctx.Jump(s.stepIsolationNegotiation)
 	}
 
-	s.semaphoreOrdered = semaphoreOrdered
-	s.semaphoreUnordered = semaphoreUnordered
+	switch objectState {
+	case object.Unknown:
+		panic(throw.Impossible())
+	case object.Missing:
+		s.prepareExecutionError(throw.E("object does not exist", struct {
+			ObjectReference string
+			State           object.State
+		}{
+			ObjectReference: s.execution.Object.String(),
+			State:           objectState,
+		}))
+		return ctx.Jump(s.stepSendCallResult)
+	case object.Inactive:
+		panic(throw.NotImplemented())
+	case object.HasState:
+		// ok
+	}
 
-	s.execution.ObjectDescriptor = objectDescriptor
+	return ctx.Jump(s.stepWaitPendingConstructorFinished)
```

Comment:
I don't like the idea of always going to stepWaitPendingConstructorFinished, even if object is not pending consturctor. The semantics of this is just wrong. I'd rather have
```
if s.pendingConstructorFinished.IsZero() {
  return ctx.Jump(s.stepWaitPendingConstructorFinished)
}
return ctx.Jump(s.stepIsolationNegotiation)
```

So reading the log will clearly distinguish pending constructors from not pending constructors.

Reply:
fixed